pkg/vuln/nvdapi/detector.go

```diff
@@ -0,0 +1,434 @@
+	// TODO: Should we only use the _first_ configuration, like secfixes-tracker does?
+
+	// TODO: Consider dropping CVE matches for really old CVEs (because the
+	//  data tends to be lower quality).
+
+	if cve == nil {
+		return nil, errors.New("input CVE was nil")
+	}
+
+	for _, configuration := range cve.Configurations {
+		if configuration.Operator == "AND" {
+			// For now, let's ignore AND-ed nodes. They usually mean that something is
+			// running on a specific platform, which usually means it's not the software
+			// we're looking for. E.g. https://nvd.nist.gov/vuln/detail/CVE-2009-3654.
+
+			continue
+		}
+
+		for _, node := range configuration.Nodes {
+			if node.Negate {
+				// TODO: Decide what to do with this case and why.
+
+				continue
+			}
+
+			for _, cpeMatch := range node.CpeMatch {
+				doCPEsMatch, err := cpeStringsMatch(requestCPE, cpeMatch.Criteria)
+				if err != nil {
+					return nil, err
+				}
+				if !doCPEsMatch {
+					continue
+				}
+
+				if !cpeMatch.Vulnerable {
+					continue
+				}
+
+				// TODO: Should we discard CpeMatches with no version data whatsoever?
+
+				vr, err := convertCpeMatchToVersionRange(cpeMatch)
+				if err != nil {
+					return nil, err
+				}
+
+				m := vuln.Match{
+					Package: vuln.Package{
+						Name: packageName,
+					},
+					CPE: vuln.CPE{
+						URI:          cpeMatch.Criteria,
+						VersionRange: vr,
+					},
+					Vulnerability: vuln.Vulnerability{
+						ID:  cve.ID,
+						URL: fmt.Sprintf("https://nvd.nist.gov/vuln/detail/%s", cve.ID),
+					},
+				}
+
+				return &m, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func cpeStringsMatch(requestCPE, responseCPE string) (bool, error) {
+	req, err := wfn.Parse(requestCPE)
+	if err != nil {
+		return false, fmt.Errorf("cannot compare CPEs: %w", err)
+	}
+	resp, err := wfn.Parse(responseCPE)
+	if err != nil {
+		return false, fmt.Errorf("cannot compare CPEs: %w", err)
+	}
+
+	// Return "false" when the request CPE had no TargetSW set, but the response CPE
+	// did have one set. This case usually means that the response CVE is for a
+	// plugin or language-specific package. (For example, a Jenkins plugin.)
+	if req.TargetSW == wfn.Any && resp.TargetSW != wfn.Any {
+		return false, nil
+	}
+
+	return wfn.Match(req, resp), nil
+}
+
+var ErrRateLimited = errors.New("we've been rate limited by NVD! 🙊")
+
+func (s *Detector) doSearch(ctx context.Context, cpe string) ([]Cve, error) {
+	err := s.rateLimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Deal with pages (not urgent because the default page size is 2,000
+	//  CVEs, and we're searching for single packages at a time.)
+
+	// TODO: Consider using an in-memory cache of API responses, keyed by the
+	//  requested CPE. This would save time on redundant API requests, like searching
+	//  for '...*:go...' multiple times because we've pruned versions from multiple,
+	//  related packages like 'go-1.18', 'go-1.19', and 'go-1.20'.
```

Comment:
❤️ This will be especially important in the private packages repo where we expect to have a lot of old versions in play at a time

Reply:
💯